vertexai/lib/main.dart

```diff
@@ -0,0 +1,548 @@
+// limitations under the License.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_markdown/flutter_markdown.dart';
+import 'package:firebase_vertexai/firebase_vertexai.dart';
+import 'package:firebase_core/firebase_core.dart';
+
+void main() {
+  runApp(const GenerativeAISample());
+}
+
+class GenerativeAISample extends StatelessWidget {
+  const GenerativeAISample({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter + Vertex AI',
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(
+          brightness: Brightness.dark,
+          seedColor: const Color.fromARGB(255, 171, 222, 244),
+        ),
+        useMaterial3: true,
+      ),
+      home: const ChatScreen(title: 'Flutter + Vertex AI'),
+    );
+  }
+}
+
+class ChatScreen extends StatefulWidget {
+  const ChatScreen({super.key, required this.title});
+
+  final String title;
+
+  @override
+  State<ChatScreen> createState() => _ChatScreenState();
+}
+
+class _ChatScreenState extends State<ChatScreen> {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(widget.title),
+      ),
+      body: const ChatWidget(),
+    );
+  }
+}
+
+class ChatWidget extends StatefulWidget {
+  const ChatWidget({
+    super.key,
+  });
+
+  @override
+  State<ChatWidget> createState() => _ChatWidgetState();
+}
+
+class _ChatWidgetState extends State<ChatWidget> {
+  late final GenerativeModel _model;
+  late final GenerativeModel _functionCallModel;
+  late final ChatSession _chat;
+  final ScrollController _scrollController = ScrollController();
+  final TextEditingController _textController = TextEditingController();
+  final FocusNode _textFieldFocus = FocusNode();
+  final List<({Image? image, String? text, bool fromUser})> _generatedContent =
+      <({Image? image, String? text, bool fromUser})>[];
+  bool _loading = false;
+
+  @override
+  void initState() {
+    super.initState();
+
+    initFirebase().then((value) {
+      _model = FirebaseVertexAI.instance.generativeModel(
+        model: 'gemini-1.5-flash-preview-0514',
+      );
+      _functionCallModel = FirebaseVertexAI.instance.generativeModel(
+        model: 'gemini-1.5-flash-preview-0514',
+        tools: [
+          Tool(functionDeclarations: [exchangeRateTool]),
+        ],
+      );
+      _chat = _model.startChat();
+    });
+  }
+
+  Future<Map<String, Object?>> findExchangeRate(
+    Map<String, Object?> arguments,
+  ) async =>
+      // This hypothetical API returns a JSON such as:
+      // {"base":"USD","date":"2024-04-17","rates":{"SEK": 0.091}}
+      {
+        'date': arguments['currencyDate'],
+        'base': arguments['currencyFrom'],
+        'rates': <String, Object?>{arguments['currencyTo']! as String: 0.091},
+      };
+
+  final exchangeRateTool = FunctionDeclaration(
```

Comment:
We may need to change this example. cc @rachelsaunders

Reply:
I double checked both generative function calling github sample as well as the documentation for it, seems it's not aligned. We will update this sample in the snippet project

Reply:
We'll soon be updating the snippets and overall "narrative" for the function calling page for both the Google AI SDKs and the Vertex AI for Firebase SDKs.  So, I wouldn't bother putting the _current_ snippets into GitHub just yet.

Reply:
sg, let me remove that part for now and the get the PR merged.

Reply:
done